Framework/MDAlgorithms/src/IntegrateEllipsoidsTwoStep.cpp

```diff
@@ -0,0 +1,553 @@
+    PARALLEL_CRITICAL(addEvents) { integrator.addEvents(qList, hkl_integ); }
+
+    prog.report();
+    PARALLEL_END_INTERUPT_REGION
+  } // end of loop over spectra
+  PARALLEL_CHECK_INTERUPT_REGION
+}
+
+/**
+ * @brief qListFromHistoWS creates qlist from input workspaces of type
+ * Workspace2D
+ * @param integrator : itegrator object on which qlists are accumulated
+ * @param prog : progress object
+ * @param wksp : input Workspace2D
+ * @param UBinv : inverse of UB matrix
+ * @param hkl_integ ; boolean for integrating in HKL space
+ */
+void IntegrateEllipsoidsTwoStep::qListFromHistoWS(Integrate3DEvents &integrator,
+                                                  Progress &prog,
+                                                  Workspace2D_sptr &wksp,
+                                                  DblMatrix const &UBinv,
+                                                  bool hkl_integ) {
+
+  // loop through the eventlists
+  const std::string ELASTIC("Elastic");
+  /// Only convert to Q-vector.
+  const std::string Q3D("Q3D");
+  const std::size_t DIMS(3);
+
+  MDWSDescription m_targWSDescr;
+  m_targWSDescr.setMinMax(std::vector<double>(3, -2000.),
+                          std::vector<double>(3, 2000.));
+  m_targWSDescr.buildFromMatrixWS(wksp, Q3D, ELASTIC);
+  m_targWSDescr.setLorentsCorr(false);
+
+  // generate the detectors table
+  Mantid::API::Algorithm_sptr childAlg = createChildAlgorithm(
+      "PreprocessDetectorsToMD", 0.,
+      .5); // HACK. soft dependency on non-dependent package.
+  childAlg->setProperty("InputWorkspace", wksp);
+  childAlg->executeAsChildAlg();
+
+  DataObjects::TableWorkspace_sptr table =
+      childAlg->getProperty("OutputWorkspace");
+  if (!table)
+    throw(std::runtime_error(
+        "Can not retrieve results of \"PreprocessDetectorsToMD\""));
+  else
+    m_targWSDescr.m_PreprDetTable = table;
+
+  int numSpectra = static_cast<int>(wksp->getNumberHistograms());
+  PARALLEL_FOR_IF(Kernel::threadSafe(*wksp))
+  for (int i = 0; i < numSpectra; ++i) {
+    PARALLEL_START_INTERUPT_REGION
+
+    // units conversion helper
+    UnitsConversionHelper unitConverter;
+    unitConverter.initialize(m_targWSDescr, "Momentum");
+
+    // initialize the MD coordinates conversion class
+    MDTransfQ3D qConverter;
+    qConverter.initialize(m_targWSDescr);
+
+    // get tof and counts
+    const auto &xVals = wksp->points(i);
+    const auto &yVals = wksp->counts(i);
+
+    // update which pixel is being converted
+    std::vector<Mantid::coord_t> locCoord(DIMS, 0.);
+    unitConverter.updateConversion(i);
+    qConverter.calcYDepCoordinates(locCoord, i);
+
+    // loop over the events
+    double signal(1.);  // ignorable garbage
+    double errorSq(1.); // ignorable garbage
+
+    std::vector<std::pair<double, V3D>> qList;
+
+    for (size_t j = 0; j < yVals.size(); ++j) {
+      const double &yVal = yVals[j];
+      if (yVal > 0) // TODO, is this condition right?
```

Comment:
Again a copy paste issue. I believe the answer is yes. If you have no counts then there are no events to be added for this bin. I do not know what else you would do.